warehouse/warehouse.go

```diff
@@ -1066,6 +1093,239 @@ func processHandler(w http.ResponseWriter, r *http.Request) {
 	recordStagedRowsStat(stagingFile.TotalEvents, stagingFile.BatchDestination.Destination.DestinationDefinition.Name, stagingFile.BatchDestination.Destination.ID, stagingFile.BatchDestination.Source.Name, stagingFile.BatchDestination.Destination.Name)
 }
 
+func pendingEventsHandler(w http.ResponseWriter, r *http.Request) {
+	// TODO : respond with errors in a common way
+	pkgLogger.LogRequest(r)
+
+	// read body
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		pkgLogger.Errorf("[WH]: Error reading body: %v", err)
+		http.Error(w, "can't read body", http.StatusBadRequest)
+		return
+	}
+	defer r.Body.Close()
+
+	// unmarshall body
+	var pendingEventsReq warehouseutils.PendingEventsRequestT
+	err = json.Unmarshal(body, &pendingEventsReq)
+	if err != nil {
+		pkgLogger.Errorf("[WH]: Error unmarshalling body: %v", err)
+		http.Error(w, "can't unmarshall body", http.StatusBadRequest)
+		return
+	}
+
+	sourceID := pendingEventsReq.SourceID
+
+	// return error if source id is empty
+	if sourceID == "" {
+		pkgLogger.Errorf("[WH]: pending-events:  Empty source id")
+		http.Error(w, "empty source id", http.StatusBadRequest)
+		return
+	}
+
+	pendingEvents := false
+	var pendingStagingFileCount int64
+	var pendingUploadCount int64
+
+	// check whether there are any pending staging files or uploads for the given source id
+	// get pending staging files
+	pendingStagingFileCount, err = getPendingStagingFileCount(sourceID)
+	if err != nil {
+		err := fmt.Errorf("Error getting pending staging file count : %v", err)
+		pkgLogger.Errorf("[WH]: %v", err)
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	// get pending uploads only if there are no pending staging files
+	if pendingStagingFileCount == 0 {
+		pendingUploadCount, err = getPendingUploadCount(sourceID)
+		if err != nil {
+			err := fmt.Errorf("Error getting pending uploads : %v", err)
+			pkgLogger.Errorf("[WH]: %v", err)
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+	}
+
+	// if there are any pending staging files or uploads, set pending events as true
+	if (pendingStagingFileCount + pendingUploadCount) > int64(0) {
+		pendingEvents = true
+	}
+
+	// trigger upload if there are pending events
+	if pendingEvents {
```

Comment:
let us take a qquery param along with the endpoint `/v1/warehouse/pending-events?triggerPending=true` and only trigger if true so that we can also use the endpoint to just get pending events without triggering. This can be used later for features like disabling Sync Now button in case there is a pending upload going on etc.